src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,337 @@
+
+/--The setoid defined by the double_coset relation-/
+def setoid (H K : subgroup G) : setoid G :=
+⟨double_coset_rel H K, rel_is_equiv⟩
+
+/--Quotient of `G` by the double coset relation, i.e. `H \ G / K`-/
+def quotient (H K : subgroup G) : Type* := quotient (setoid H K)
+
+/--The double_coset as an element of `set α` corresponding to `s a t` -/
+def doset (s t : set α) (a : α) : set α := { b : α | ∃ x : (s × t), b = x.1 * a * x.2 }
+
+@[simp]
+lemma doset_mem (s t : set α) (a b : α) :
+  b ∈ (doset s t a) ↔  ∃ x : (s × t), b = x.1 * a * x.2 :=iff.rfl
+
+lemma sub_doset  (H K : subgroup G) (a b : G) (hb : b ∈ doset H.1 K a) :
+  (doset H.1 K b) ⊆ (doset H K a) :=
+begin
+  intro x,
+  simp only [and_imp, exists_prop, forall_exists_index, set_coe.exists, subgroup.mem_carrier,
+  set_like.mem_coe, doset_mem,
+  subgroup.coe_mk] at *,
+  intros g hg,
+  cases hb,
+  rw hb_h at hg,
+  use ⟨g.1 * hb_w.1, hb_w.2 * g.2⟩,
+  simp [hg],
+  simp_rw ← mul_assoc,
+end
+
+--Is this already somewhere? I'd rather not have to prove it
+lemma auxeq : ∀ (a b c d e f : G), a * b * c = d * e * f ↔ b = a⁻¹ * d * e * f * c⁻¹ :=
+begin
+  intros a b c d e f,
+  split; intro h,
+  { assoc_rw h.symm,
+    group, },
+  { rw h,
+    group, },
+end
+
+lemma disjoint_sub (H K : subgroup G) (a b : G) (h : ¬ disjoint (doset H.1 K a ) (doset H K b)) :
+  b ∈ doset H.1 K a :=
+begin
+  rw set.not_disjoint_iff at h,
+  simp only [exists_prop, set_coe.exists, doset_mem, subgroup.mem_carrier, set_like.mem_coe,
+    subgroup.coe_mk] at *,
+  rcases h with ⟨x, ⟨hx, hxx⟩, hk, hhk⟩ ,
+  use ( ⟨hk.1⁻¹ * hx.1, hx.2 * hk.2⁻¹⟩ : H × K),
+  rw hxx at hhk,
+  simp only [subgroup.coe_inv, subgroup.coe_mul],
+  simp_rw  ← mul_assoc,
+  rw ← auxeq,
+  apply hhk.symm,
+end
+
+lemma disjoint_doset  (H K : subgroup G) (a b : G) (h: ¬ disjoint (doset H.1 K a) (doset H K b)) :
+   doset H.1 K a = doset H K b :=
+begin
+  have hb :  b ∈ (doset H.1 K a), by {apply disjoint_sub _ _ _ _ h},
+  rw disjoint.comm at h,
+  have ha :  a ∈ (doset H.1 K b), by {apply disjoint_sub _ _ _ _ h},
+  rw set.subset.antisymm_iff,
+  split,
+  apply sub_doset H K b a ha,
+  apply sub_doset H K a b hb,
+end
+
+/--Create a doset out of an element of `H \ G / K`-/
+def quot_to_doset (H K : subgroup G) (q : quotient H K ) : set G := (doset H.1 K q.out')
+
+/--Map from `G` to `H \ G / K`-/
+abbreviation mk (H K : subgroup G) (a : G) : quotient H K  :=
+quotient.mk' a
+
+instance (H K : subgroup G) : inhabited (quotient H K) := ⟨(mk H K (1 : G) : quotient H K)⟩
+
+lemma eq (H K : subgroup G) (a b : G): mk H K a = mk H K b ↔ ∃ (h ∈ H) (k ∈ K), b = h * a * k :=
+begin
+rw quotient.eq',
+apply  (rel_iff H K a b),
+end
+
+lemma out_eq' (H K : subgroup G) (q : quotient H K) : mk H K q.out' = q :=
+quotient.out_eq' q
+
+lemma mk_out'_eq_mul (H K : subgroup G)  (g : G) :
+  ∃ (h k : G), (h ∈ H) ∧ (k ∈ K) ∧ (mk H K g : quotient H K).out' = h * g * k :=
+begin
+  have := eq  H K (mk H K g :  quotient H K).out' g,
```

Comment:
You have some weird double spaces lying around throughout the file. Can you fix these?

Reply:
Done